### PR TITLE
team-management: generate manifests automatically

### PR DESCRIPTION
**Pull request by yokaze**
This PR introduces jsonnet template for files updated in team management procedure.
The location of each libsonnet files represents the corresponding YAML path.
The files under `team-management/template/team-management` is arranged:
```
├── base
│   ├── common
│   │   ├── main.libsonnet
│   │   └── project.libsonnet
│   ├── kustomization.libsonnet
│   └── team-template
│       ├── kustomization.libsonnet
│       ├── main.libsonnet
│       ├── namespace.libsonnet
│       └── project.libsonnet
└── main.libsonnet
```
I plan to make `team-management/template/argocd-config` and `team-management/template/teleport` in future PRs.

`main.libsonnet` is the representative for all the configuration under its territory.
It generates key-value pairs representing the manifest path and its content. It looks like:
```
{
  "a.yaml": <json content>,
  "dir/b.yaml": <json content>
}
```
`make all` transforms it into separate YAML files.

Most of the libsonnet files consumes `teams`, `team-namespaces` and `all-tenant-namespaces` as their arguments.
The configuration is provided by `settings.json` at the root directory of the template system (`team-management/template`).
Direct modification of the file results in new teams and namespaces.
We can do it easier using:
```
$ ./add_team.sh new_team
$ ./add_namespace.sh team new_namespace
then
$ make all
```

`make format` formats all the json and the jsonnet files under the template system.

CI checks:
- all the files are formatted appropriately
- the manifests are generated from the current `settings.json`.

Signed-off-by: Daichi Sakaue <daichi-sakaue@cybozu.co.jp>

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -157,17 +157,31 @@ jobs:
...
+            make setup SUDO=
+      - run:
+          name: Setup test tools
           command: |
             cd test
             make setup SUDO=
+      - save_cache:
+          key: team-management-download-{{ checksum "CACHE_VERSION" }}-{{ checksum "team-management/template/Makefile" }}
+          paths:
+            - team-management/template/download
```
Comment (yokaze) [inline .circleci/config.yml:177]: Merge two downloaded files into `/download`.

File: .gitignore [hunk 2]
```diff
@@ -11,9 +11,14 @@
...
 *.out
 
+# Ignore CI files
+/CACHE_VERSION
+
 # Ignore downloaded files
 /logging/bin
 /rook/bin
+/team-management/template/bin
+/team-management/template/download
```
Comment (yokaze) [inline .gitignore:21]: Move to `/download`

File: team-management/template/Makefile [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+YQ_VERSION := 4.6.3
+AUTOGEN_COMMENT := This file is automatically generated from a template. Please do not edit.
+
+ROOT_DIR := $(shell git rev-parse --show-toplevel)
+BINDIR := $(abspath $(CURDIR)/bin)
+JSONNET := $(BINDIR)/jsonnet
+JSONNETFMT := $(BINDIR)/jsonnetfmt
+YQ := $(BINDIR)/yq
+
+DOWNLOAD_DIR := $(abspath $(CURDIR)/download)
```
Comment (yokaze) [inline team-management/template/Makefile:11]: Move to `/download`

File: team-management/template/Makefile [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+WGET := wget --retry-connrefused --no-verbose
+WGET_PATH := $(shell which wget || echo wget-install)
+
+.PHONY: all
+all: team-management
+
+.PHONY: team-management
+team-management:
+	mv ../base/ept/elastic-serviceaccount.yaml /tmp/ept-elastic-serviceaccount.yaml
```
Comment (yokaze) [inline team-management/template/Makefile:23]:
```
# Cleanup existing files
```

File: team-management/template/Makefile [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+YQ_DLPATH := $(DOWNLOAD_DIR)/yq-v$(YQ_VERSION).tar.gz
+
+WGET := wget --retry-connrefused --no-verbose
+WGET_PATH := $(shell which wget || echo wget-install)
+
+.PHONY: all
+all: team-management
+
+.PHONY: team-management
+team-management:
```
Comment (yokaze) [inline team-management/template/Makefile:22]: # Cleanup existing files

File: team-management/template/Makefile (hunk 4 above)
Comment (yokaze) [inline team-management/template/Makefile:23]: Generate `ServiceAccount` for all the teams and the namespaces.

File: team-management/template/team-management/base/team-template/kustomization.libsonnet [hunk 6]
```diff
@@ -0,0 +1,14 @@
+function(team, namespaces) [
+  {
+    apiVersion: 'kustomize.config.k8s.io/v1beta1',
+    kind: 'Kustomization',
+    resources: std.set(
+      ['project.yaml'] +
+      std.map(function(x) x + '.yaml', namespaces) +
+      if team == 'ept' || team == 'maneki' then ['elastic-serviceaccount.yaml'] else []
```
Comment (yokaze) [inline team-management/template/team-management/base/team-template/kustomization.libsonnet:8]: Provide `elastic-serviceaccount` for all teams.

File: team-management/template/team-management/base/team-template/namespace.libsonnet [hunk 7]
```diff
@@ -0,0 +1,47 @@
+function(team, namespace) [
+  {
+    apiVersion: 'v1',
+    kind: 'Namespace',
+    metadata: {
+      name: namespace,
+      [if team == 'maneki' && namespace == 'maneki' then 'labels']: {
```
Comment (yokaze) [inline team-management/template/team-management/base/team-template/namespace.libsonnet:7]:
```
// This label may be unused.
```

File: team-management/template/team-management/base/team-template/namespace.libsonnet [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        'app.kubernetes.io/name': 'maneki',
+      },
+    },
+  },
+  {
+    apiVersion: 'rbac.authorization.k8s.io/v1',
+    kind: 'RoleBinding',
+    metadata: {
+      name: if team == 'maneki' && namespace == 'app-monitoring' then 'monitoring-role-binding' else team + '-role-binding',
```
Comment (yokaze) [inline team-management/template/team-management/base/team-template/namespace.libsonnet:16]: Remove if branch.

**Review by zoetrope — CHANGES_REQUESTED** [review]
reviewed

**Review by zoetrope — APPROVED** [review]
LGTM

**Pull request merged**